scene/main/scene_tree.cpp

```diff
@@ -551,6 +552,9 @@ void SceneTree::quit(int p_exit_code) {
 		// Override the exit code if a positive argument is given (the default is `-1`).
 		// This is a shorthand for calling `set_exit_code()` on the OS singleton then quitting.
 		OS::get_singleton()->set_exit_code(p_exit_code);
+	} else if (!OS::get_singleton()->is_custom_exit_code()) {
```

Comment:
I'm not sure I like this `is_custom_exit_code` logic.

It would expect that when we call `quit()`, either we specify an exit code, or we default to `EXIT_SUCCESS`.

If we currently have situations where we `quit()` on failure with a non-zero exit code, the code that calls `quit()` should be fixed to use `quit(EXIT_FAILURE)`.

Reply:
I was thinking about this PR the other day, I think most of the complexity comes from `set_exit_code` which feels like a legacy artifact now we can provide the exit code to `quit()` (passing exit code while calling the exit function is what is done in basically every language)

So if we take advantage of Godot 4.0, we can ditch `set_exit_code` and have `quit()` with `EXIT_SUCCESS` as default value. The `is_custom_exit_code ` stuff is not needed anymore (`exit_code` gets initialized with `EXIT_FAILURE`, and is overwritten when calling `quit()`) ;-)

Reply:
I've updated the PR with 614d890b8944cab2819982039bf34461b8e8492a in this view